File: src/test/java/net/pterodactylus/sone/web/CreateAlbumPageTest.java
package net.pterodactylus.sone.web;

import static net.pterodactylus.sone.web.WebTestUtils.redirectsTo;
import static net.pterodactylus.util.web.Method.POST;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.mockito.Answers.RETURNS_DEEP_STUBS;
import static org.mockito.Answers.RETURNS_SELF;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import net.pterodactylus.sone.data.Album;
import net.pterodactylus.sone.data.Album.Modifier;
import net.pterodactylus.sone.data.Album.Modifier.AlbumTitleMustNotBeEmpty;
import net.pterodactylus.sone.test.Dirty;

import org.junit.Test;

/**
 * Unit test for {@link CreateAlbumPage}.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class CreateAlbumPageTest extends WebPageTest {

	private final CreateAlbumPage page = new CreateAlbumPage(template, webInterface);

	@Test
	public void pageReturnsCorrectPath() {
		assertThat(page.getPath(), is("createAlbum.html"));
	}

	@Test
	public void getRequestShowsTemplate() throws Exception {
		page.processTemplate(freenetRequest, templateContext);
	}

	@Test
	public void missingNameResultsInAttributeSetInTemplateContext() throws Exception {
		request("", POST);
		page.processTemplate(freenetRequest, templateContext);
		assertThat(templateContext.get("nameMissing"), is((Object) true));
	}

	@Test
	public void titleAndDescriptionAreSetCorrectlyOnTheAlbum() throws Exception {
		request("", POST);
		Album parentAlbum = createAlbum("parent-id");
		when(core.getAlbum("parent-id")).thenReturn(parentAlbum);
		Album newAlbum = createAlbum("album-id");
		when(core.createAlbum(currentSone, parentAlbum)).thenReturn(newAlbum);
		addHttpRequestParameter("name", "new name");
		addHttpRequestParameter("description", "new description");
		addHttpRequestParameter("parent", "parent-id");
		expectedException.expect(redirectsTo("imageBrowser.html?album=album-id"));
		try {
			page.processTemplate(freenetRequest, templateContext);
		} finally {
			verify(newAlbum).modify();
			verify(newAlbum.modify()).setTitle("new name");
			verify(newAlbum.modify()).setDescription("new description");
			verify(newAlbum.modify()).update();
		}
	}

	private Album createAlbum(String albumId) {
		Album newAlbum = mock(Album.class, RETURNS_DEEP_STUBS);
		when(newAlbum.getId()).thenReturn(albumId);
		Modifier albumModifier = mock(Modifier.class, RETURNS_SELF);
		when(newAlbum.modify()).thenReturn(albumModifier);
		when(albumModifier.update()).thenReturn(newAlbum);
		return newAlbum;
	}

	@Test
	public void rootAlbumIsUsedIfNoParentIsSpecified() throws Exception {
		request("", POST);
		Album parentAlbum = createAlbum("root-id");
		when(currentSone.getRootAlbum()).thenReturn(parentAlbum);
		Album newAlbum = createAlbum("album-id");
		when(core.createAlbum(currentSone, parentAlbum)).thenReturn(newAlbum);
		addHttpRequestParameter("name", "new name");
		addHttpRequestParameter("description", "new description");
		expectedException.expect(redirectsTo("imageBrowser.html?album=album-id"));
		page.processTemplate(freenetRequest, templateContext);
	}

	@Test
	@Dirty("that exception can never happen")
	public void emptyAlbumTitleRedirectsToErrorPage() throws Exception {
		request("", POST);
		Album parentAlbum = createAlbum("root-id");
		when(currentSone.getRootAlbum()).thenReturn(parentAlbum);
		Album newAlbum = createAlbum("album-id");
		when(core.createAlbum(currentSone, parentAlbum)).thenReturn(newAlbum);
		when(newAlbum.modify().update()).thenThrow(AlbumTitleMustNotBeEmpty.class);
		addHttpRequestParameter("name", "new name");
		addHttpRequestParameter("description", "new description");
		expectedException.expect(redirectsTo("emptyAlbumTitle.html"));
		page.processTemplate(freenetRequest, templateContext);
	}

}

File: src/test/java/net/pterodactylus/sone/web/CreateAlbumPageTest.kt
package net.pterodactylus.sone.web

import net.pterodactylus.sone.data.Album
import net.pterodactylus.sone.data.Album.Modifier.AlbumTitleMustNotBeEmpty
import net.pterodactylus.sone.test.deepMock
import net.pterodactylus.sone.test.selfMock
import net.pterodactylus.sone.test.whenever
import net.pterodactylus.util.web.Method.POST
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.verify

/**
 * Unit test for [CreateAlbumPage].
 */
class CreateAlbumPageTest: WebPageTest() {

	private val page = CreateAlbumPage(template, webInterface)

	override fun getPage() = page

	private val parentAlbum = createAlbum("parent-id")
	private val newAlbum = createAlbum("album-id")

	@Before
	fun setupAlbums() {
		whenever(core.createAlbum(currentSone, parentAlbum)).thenReturn(newAlbum)
		whenever(currentSone.rootAlbum).thenReturn(parentAlbum)
	}

	@Test
	fun `page returns correct path`() {
		assertThat(page.path, equalTo("createAlbum.html"))
	}

	@Test
	fun `get request shows template`() {
		page.processTemplate(freenetRequest, templateContext)
	}

	@Test
	fun `missing name results in attribute being set in template context`() {
		request("", POST)
		page.processTemplate(freenetRequest, templateContext)
		assertThat(templateContext["nameMissing"], equalTo<Any>(true))
	}

	private fun createAlbum(albumId: String) = deepMock<Album>().apply {
		whenever(id).thenReturn(albumId)
		selfMock<Album.Modifier>().let { modifier ->
			whenever(modifier.update()).thenReturn(this@apply)
			whenever(this@apply.modify()).thenReturn(modifier)
		}
	}

	@Test
	fun `title and description are set correctly on the album`() {
		request("", POST)
		addAlbum("parent-id", parentAlbum)
		addHttpRequestParameter("name", "new name")
		addHttpRequestParameter("description", "new description")
		addHttpRequestParameter("parent", "parent-id")
		verifyRedirect("imageBrowser.html?album=album-id") {
			verify(newAlbum).modify()
			verify(newAlbum.modify()).setTitle("new name")
			verify(newAlbum.modify()).setDescription("new description")
			verify(newAlbum.modify()).update()
			verify(core).touchConfiguration()
		}
	}

	@Test
	fun `root album is used if no parent is specified`() {
		request("", POST)
		addHttpRequestParameter("name", "new name")
		addHttpRequestParameter("description", "new description")
		verifyRedirect("imageBrowser.html?album=album-id")
	}

	@Test
	fun `empty album title redirects to error page`() {
		request("", POST)
		whenever(newAlbum.modify().update()).thenThrow(AlbumTitleMustNotBeEmpty::class.java)
		addHttpRequestParameter("name", "new name")
		addHttpRequestParameter("description", "new description")
		verifyRedirect("emptyAlbumTitle.html")
	}

	@Test
	fun `album description is filtered`() {
		request("", POST)
		addHttpRequestParameter("name", "new name")
		addHttpRequestParameter("description", "new http://localhost:12345/KSK@foo description")
		addHttpRequestHeader("Host", "localhost:12345")
		verifyRedirect("imageBrowser.html?album=album-id") {
			verify(newAlbum.modify()).setDescription("new KSK@foo description")
		}
	}

}
